core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -165,7 +232,13 @@
       MasterUtils.createMasters(mJournalSystem, mRegistry, mSafeModeManager);
 
       // Create config checker
+      mMasterConfigChecker = new ServerConfigurationChecker();
       mWorkerConfigChecker = new ServerConfigurationChecker();
+      mClock = new SystemClock();
+      mExecutorService = ExecutorServiceFactories
+          .fixedThreadPoolExecutorServiceFactory(Constants.META_MASTER_NAME, 2).create();
+      mMetaMasterClient = new MetaMasterMasterClient(MasterClientConfig.defaults());
+      setMasterIdAndHostname();
```

Comment:
Does the primary heartbeat to itself?

Reply:
In the AlluxioMasterProcess.startMasters(), if a master is leader master, it will setup lostMasterDectationHeartbeatExecutor to detect lost standby masters. If a master is not a leader master, it will setup its MetaMasterSync to getMasterId, register and heartbeat with the leader master.
``` java
    try {
      mWorkerConfigChecker.reset();
      if (isLeader) {
        mSafeModeManager.notifyPrimaryMasterStarted();

        //  The service that detects lost standby master nodes
        mExecutorService.submit(new HeartbeatThread(
            HeartbeatContext.MASTER_LOST_MASTER_DETECTION,
            new LostMasterDetectionHeartbeatExecutor(),
            (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      } else {
        // Standby master should setup MetaMasterSync to communicate with the leader master
        mMetaMasterSync = new MetaMasterSync(mMasterId, mMasterHostname, mMetaMasterClient);
        mExecutorService.submit(new HeartbeatThread(HeartbeatContext.MASTER_SYNC, mMetaMasterSync,
            (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      }
```
So primary will not send heartbeat/getMasterId/register to itself